README.md
# Teamboard API

`teamboard-api` is the API Teamboard uses.

## Dependencies

The API depends on having [MongoDB](http://mongodb.org) and
[Redis](http://redis.io) up and running.

## Installation
```
npm install N4SJAMK/teamboard-api
```

### Setup

Set `NODE_ENV` to `development` for local development, assuming you have
development dependencies installed locally. If you are running in `production`
environment, you must specify the required environmental variables. There is
also a `test` environment that is used for test runs.

- `MONGODB_URL` is actually the connection string passed to `mongoose`, for
  example: `mongodb://localhost:27017/teamboard-dev`.
- `REDIS_HOST` and `REDIS_PORT` corresponding to your running `redis-server`.
- `TOKEN_SECRET` is the secret used with `jsonwebtoken`.

You can also specify the `PORT` variable to change the default port of `9002`.

### Running

Once you've set the required variables, run:
```
npm start
```

## Testing
```
npm test
```

## API Methods

All the API methods except `/auth/login` and `/auth/register` expect that the
`Authorization` header is set to `Bearer access-token` where `access-token` is
the `x-access-token` received from providing correct credentials. Failure to
provide a proper `access-token` will be met with a swift 401.

### Auth

**GET /auth**
```
returns:
{
  "id":       "id"
  "type":     "user  | guest",
  "username": "email | username",
}
```
Returns the `user` object tied to `access-token` passed in `Authorization`
header. If the user's `type` is `guest`, `access` field will also be present in
the response containing the `id` of the `board` the token is tied to.


**POST /auth/login**
```
payload:
{
  "email":    "email",
  "password": "password"
}
```
```
returns:
{
  "id":       "id",
  "type":     "user",
  "username": "email"
}
```
Exchanges the given credentials for an `access-token`. The `access-token` is
returned in the `x-access-token` header. The `type` of the token received from
this method is always `user`.


**POST /auth/register**
```
payload:
{
  "email":    "email",
  "password": "password"
}
```
```
returns:
{
  "id":    "id",
  "email": "email"
}
```
Creates a new user.


**POST /auth/logout**

Removes the given token from the matching user, essentially logging the user
out.


### Board

**GET /boards**
```
returns:
[
  {
    "id": "id",

    "name":        "name",
    "description": "description",

    "createdBy":  "user.id",
    "accessCode": "",

    "background": "none",
    "size": {
      "width":  8,
      "height": 8
    }
  }
]
```
Returns an array of `boards` that are created by the `user` identified by the
given `access-token`.

**GET /boards/:id**

Returns the `board` object specified by `:id`. See above for the description of
the `board` object.

**POST /boards**
```
payload:
{
  "name": "name"
}
```
Creates a new board with the given `name`. See above for a list of properties
that a `board` can have. Returns the created `board` object, see above for the
description of the `board` object.

**PUT /boards/:id**
```
payload:
{
  "description": "new description"
}
```
Updates the board specified by `:id`. See above for possibilities. Returns the
updated `board` object. See above for the description of the `board` object.

**DELETE /boards/:id**

Removes the board specified by `:id`. Returns the removed `board` object. See
above for the description of the `board` object.


### Tickets

**GET /boards/:id/tickets**
```
returns:
[
  {
    "id":      "id",
    "content": "content",
    "color":   "#BABABA",
    "position": {
      "x": 0, "y": 0
    }
  }
]
```
Returns an array of `ticket` objects that belong to the `board` specified by
`:id`.

**POST /boards/:id/tickets**
```
payload:
{
  "content": "content"
}
```
Creates a new `ticket`. Returns the created `ticket` object. See above for the
description of `ticket` object.

**PUT /boards/:id/tickets/:id**
```
payload:
{
  "content": "new-content"
}
```
Updates a ticket specified by `:id`. Returns the updated `ticket` object. See
above for the description of the `ticket` object.

**DELETE /boards/:id/tickets/:id**

Removes the ticket specified by `:id`. Returns the removed `ticket` object. See
above for the description of the `ticket` object.

### Ticket Comments

Comments are actually events, see below for an explanation.

**GET /boards/:id/tickets/:id/comments**
```
returns:
[ EventObject ]
```

**POST /boards/:id/tickets/:id/comments**
```
payload:
{
  'comment': '...',
}
```
```
returns:
{ EventObject }
```

### Board Events

Events are generated for various actions done via the API. Events are also
emitted through `teamboard-io`, so that connected clients can listen to them
under the key `board:event`.

Currently the possible event `types` are:
- `BOARD_CREATE`, `BOARD_EDIT`, `BOARD_REMOVE`
- `BOARD_PUBLISH`, `BOARD_UNPUBLISH`, `BOARD_GUEST_JOIN`
- `TICKET_CREATE`, `TICKET_EDIT`, `TICKET_REMOVE`, 'TICKET_COMMENT'

The returned `events` have a `data` field that is used to store additional data
regarding the `event`. For example for ticket based events this usually means
storing an `id` attribute referencing the ticket in question.

**GET /boards/:id/events**
```
returns:
[{
  'type': 'event.type',
  'board': 'board.id',
  'user': { 'id', 'type', 'username' },
  'data': { }
}]
```

### Guest Access

Guest access can be granted to a board by generating an `access-code` for it.
By using this `access-code` anonymous users can request an `access-token` that
is tied to a specific board and to the specific `access-code`.

**POST /boards/:id/access**
```
returns:
{
  "accessCode": "123ABCDE"
}
```
Generates an `access-code` for the board specified by `:id`. Successive calls
to this method will always generate a new `access-code`, thus invalidating any
`access-tokens` that were tied to the previous `access-code`.

**DELETE /boards/:id/access**

Removes the `access-code` of the board specified by `:id`, invalidating any
`access-tokens` tied to it.

**POST /boards/:id/access/:code**
```
payload:
{
  "username": "username"
}
```
```
returns:
{
  "id":       "id",
  "type":     "guest",
  "access":   "board.id",
  "username": "username"
}
```
Requests an `access-token` for the `board` specified by `:id` and authorized by
`:code`. The access token can be found in `x-access-token` after a succesful
request. The method also returns the payload encoded in the `access-token`.


CHANGELOG
Changelog
=========

v0.2.12 (2014-11-28)
--------------------

- Add support for vagrant development. [tanelih]

- Fix invalid error message with duplicate user creation. [tanelih]

v0.2.11 (2014-11-26)
--------------------

- Add documentation for 'utils/error'. [tanelih]

v0.2.4 (2014-11-26)
-------------------

- Fix issue with 'passport' not propagating errors. [tanelih]

- Fix issue with 'gulp.watch' tracking too many files. Damn you
  'node_modules'... [tanelih]

v0.2.1 (2014-11-18)
-------------------

- Fix issue with request body size being set too low. [tanelih]

v0.2.0 (2014-11-18)
-------------------

- Disable 'newrelic' by default. [Taneli Hartikainen]

- Small change to PUT /boards/:id. [Taneli Hartikainen]

- Disable 'newrelic' for test runs. [Taneli Hartikainen]

- Fix issues with updating a board, due to invalid payload. [Taneli
  Hartikainen]

- Add ignore rules to 'newrelic' configuration for 400 status codes.
  [Taneli Hartikainen]

- Fix issue with ticket updating. [Taneli Hartikainen]

- Add missing package to deps. [Taneli Hartikainen]

- Add csv-export. [Taneli Hartikainen]

- Add 'export' for 'json' format. [Taneli Hartikainen]

- Add documentation. [Taneli Hartikainen]

- Add commenting. [Taneli Hartikainen]

- Prefix /api on routes. [Taneli Hartikainen]

- Fixed invalid password test in sign up. Cleaning up. [Petri
  Matilainen]

- Added password and email validation. More unit tests. [Petri
  Matilainen]

- Change static mongoose version to package.json. [spherelot]

- Add 'board' attribute to emitted events. [Taneli Hartikainen]

- Add new '_members' attribute that represents what the API will be in
  the future. [Taneli Hartikainen]

- Add background and size options for board. [Taneli Hartikainen]

- Add population to PUT board. [Taneli Hartikainen]

- Add population to create-board. [Taneli Hartikainen]

- Remove stuff, add some comments. [Taneli Hartikainen]

- Remove 'err' utility function, replace with inline error handling.
  [Taneli Hartikainen]

  Reformatted codebase while at it, will add some comments and new-
  lines next.

- Remove crypt-service request retries. Remove bcrypt dependency.
  [Taneli Hartikainen]

- Remove teamboard-crypt dependency. [Taneli Hartikainen]

- Add more logging. [Taneli Hartikainen]

- Add cleaner comparison to retries. [Taneli Hartikainen]

- Add retry on request-fail. [Taneli Hartikainen]

- Tone sockets down to 128. [Taneli Hartikainen]

- Set http.globalAgent.maxSockets to 512. [Taneli Hartikainen]

- Add logging to crypt-server requests. [Taneli Hartikainen]

- Fix logging in on multiple clients destroying sessions. [Taneli
  Hartikainen]

- Add support for crypt-service. [Taneli Hartikainen]

- Add DELETE route for multiple boards, as with tickets. [Taneli
  Hartikainen]

- Fix issue with board being stuck in public state. [Taneli Hartikainen]

- Fix issues with 500 from authentication. [Taneli Hartikainen]

- Fix error handling in screenshot API. [Taneli Hartikainen]

- Add multiple ticket deletion API. Remove guest routes and resolve.
  [Taneli Hartikainen]

- Remove some debug code. [Taneli Hartikainen]

- Add implementation to screenshot api. [Taneli Hartikainen]

- Remove guest schema. Rework authentication into 'bearer' and
  'anonymous' strategies. Add stub route for screenshots. [Taneli
  Hartikainen]

- Modify emitter payload to include both ticket and user. [Taneli
  Hartikainen]

- Update README.md. [Taneli Hartikainen]

- Rename MONGODB_HOST to MONGODB_URL for clarity. [Taneli Hartikainen]

- Add installation instructions. [Taneli Hartikainen]

- Update and rename LICENCE.md to LICENSE. [Taneli Hartikainen]

  Typos!

- Create LICENCE.md. [Taneli Hartikainen]

- Initial commit. [Taneli Hartikainen]




routes/board.js
'use strict';

var _        = require('lodash');
var express  = require('express');
var mongoose = require('mongoose');
var Promise  = require('promise');
var request  = require('request');

var utils      = require('../utils');
var config     = require('../config');
var middleware = require('../middleware');

var Event  = mongoose.model('event');
var Board  = mongoose.model('board');
var Ticket = mongoose.model('ticket');
var User   = mongoose.model('user');

var Router   = express.Router();
var ObjectId = mongoose.Types.ObjectId;

var exportAs = require('../utils/export');


// automagically resolve 'id' attributes to their respective documents
Router.param('board_id',  middleware.resolve.board);
Router.param('ticket_id', middleware.resolve.ticket);

Router.route('/boards')

	/**
	 * Returns the boards that have been created by the user making the
	 * request. If the requestee is a guest, returns the board the guest has
	 * access to.
	 *
	 * returns:
	 *   An array of board objects.
	 */
	.get(middleware.authenticate('user', 'guest'))
	.get(function(req, res, next) {
		var boardQuery = null;

		if(req.user.type === 'temporary') {
			// Guests can only see the board they have access to...
			boardQuery = Board.find({ _id: req.user.access });
		}
		else {
			// Normal users see the boards they are member or admin to.
			boardQuery = Board.find({'members.id': req.user.id });

		}

		boardQuery.exec(function(err, boards) {
			if(err) {
				return next(utils.error(500, err));
			}
			return res.json(200, boards);
		});
	})

	/**
	 * Create a new board.
	 *
	 * TODO The 'background' attribute should probably be some sort of an
	 *      enumeration.
	 * TODO Validate the given payload. Can be done in the 'board' model.
	 *
	 * {
	 *   'name':        'cool-board'
	 *   'description': 'cool things only'
	 *   'background':  'none'
	 *   'size': {
	 *     'width', 'height'
	 *   }
	 * }
	 */
	.post(middleware.authenticate('user'))
	.post(function(req, res, next) {
		var payload           			 = req.body;
		payload.members                  = [];
		    //payload.members[req.user.id] = 'admin';
		payload.members.push({id: req.user.id, role: 'admin'});

		if(payload.size.height <= 0 || payload.size.width <= 0) {
			return next(utils.error(400, 'Board size must be larger than 0!'));
		}

		if(!(payload.size.height % 1 === 0) || !(payload.size.width % 1 === 0)) {
			return next(utils.error(400, 'Board size must be whole numbers!'));
		}

		if (payload.size)
		new Board(payload).save(function(err, board) {
			if(err) {
				return next(utils.error(400, err));
			}

			User.findOne({ _id: req.user.id }, function(err, doc) {
				doc.boards.push({id: board._id});
				doc.save(function(err) {
					if(err) return console.error(err);

					new Event({
						'type': 'BOARD_CREATE',
						'board': board.id,
						'user':  {
							'id':       req.user.id,
							'type':     req.user.type,
							'username': req.user.username,
						}
					}).save(function(err) {
						if(err) return console.error(err);
					});

					return res.json(201, board);
				});
			});
		});
	});


Router.route('/boards/:board_id')

	/**
	 * Get a specific board.
	 *
	 * returns:
	 *   The specified board object.
	 */
	.get(middleware.authenticate('user', 'guest'))
	.get(middleware.relation('admin', 'user', 'guest'))
	.get(function(req, res, next) {
		var boardQuery = Board.findOne({ '_id': req.resolved.board.id, 'members.id': req.user.id }).populate('members.id');

		boardQuery.exec(function(err, board) {
				if(err) {
					return next(utils.error(500, err));
				}
				return res.json(200, board);
			});
	})

	/**
	 * Update the specified board. See [POST /boards] for payload details.
	 *
	 * TODO Validate the payload. Can be done in the 'board' model.
	 *
	 * payload:
	 *   {
	 *     'name':        'new-name'
	 *     'description': 'new-info'
	 *     'background':  'new-background'
	 *     'size': {
	 *       'width', 'height'
	 *     }bout.
	 *   }
	 *
	 * returns:
	 *   The updated board object.
	 */
	.put(middleware.authenticate('user'))
	.put(middleware.relation('admin'))
	.put(function(req, res, next) {
		var old            = req.resolved.board.toObject();
		req.resolved.board = _.merge(req.resolved.board, req.body);

		var ticketWidth = 192;
		var ticketHeight = 108;

		if (!req.resolved.board) {
			return next(utils.error(404, 'Board not found!'));
		}

		if(req.resolved.board.size.height <= 0 || req.resolved.board.size.width <= 0) {
			return next(utils.error(400, 'Board size must be larger than 0!'));
		}

		if(!(req.resolved.board.size.height % 1 === 0) || !(req.resolved.board.size.width % 1 === 0)) {
			return next(utils.error(400, 'Board size must be whole numbers!'));
		}

		return req.resolved.board.save(function(err, board) {
			if(err) {
				return next(utils.error(400, err));
			}
			Board.populate(board, 'createdBy', function(err, board) {
				if(err) {
					return next(utils.error(500, err));
				}

				if(req.resolved.board.size.width < old.size.width || req.resolved.board.size.height < old.size.height){
					Ticket.find({ 'board': req.resolved.board.id,
						$or: [
								{'position.x': {$gt: (req.resolved.board.size.width * ticketWidth) - ticketWidth / 2}},
								{'position.y': {$gt: (req.resolved.board.size.height * ticketHeight) - ticketHeight / 2}}
						     ]}, function (err, tickets) {

						if(tickets.length > 0) {
							Promise.all(tickets.map(utils.ticketClamper(req.resolved.board))).then(function(){

								utils.createEditBoardEvent(req, req.resolved.board, old);
							})
						}
					});
				} else {

					utils.createEditBoardEvent(req, req.resolved.board, old);
				}

				return res.json(200, board);
			});
		});
	})

	/**
	 * Remove the specified board.
	 *
	 * returns:
	 *   The removed board object.
	 */
	.delete(middleware.authenticate('user'))
	.delete(middleware.relation('admin'))
	.delete(function(req, res, next) {
		req.resolved.board.remove(function(err) {
			if(err) {
				return next(utils.error(500, err));
			}

			new Event({
				'type':  'BOARD_REMOVE',
				'board': req.resolved.board.id,
				'user': {
					'id':       req.user.id,
					'type':     req.user.type,
					'username': req.user.username,
				}
			}).save(function(err, ev) {
				if(err) {
					return console.error(err);
				}
				utils.emitter.to(ev.board).emit('board:event', ev.toObject());
			});

			return res.json(200, req.resolved.board);
		});
	});


Router.route('/boards/:board_id/export')

	/**
	 * Export board either json, csv, plaintext or image
	 */
	.get(middleware.authenticate('user', 'guest'))
	.get(middleware.relation('admin', 'user', 'guest'))
	.get(function(req, res, next) {
		var format = req.query.format ? req.query.format : 'json';
		
		var boardQuery = Board.findById(req.resolved.board.id)
			.populate({
				'path':   'createdBy',
				'select': '-_id -__v -password -token',
			})
			.select('-_id -__v -accessCode').lean();

		boardQuery.exec(function(err, board) {
			
			if(err) {
				return next(utils.error(500, err));
			}

			var ticketQuery = Ticket.find({ 'board': req.resolved.board.id })
				.select('-_id -__v -board').lean();

			ticketQuery.exec(function(err, tickets) {
				if(err) {
					return next(utils.error(500, err));
				}

				if(board.name == '')
					board.name = 'board';
				else
					// Edit board name incase name is not valid filename
					board.name = utils.sanitize(board.name,'');

				if(format == 'csv') {
					return res.attachment(board.name + '.csv').send(200, exportAs.generateCSV(board, tickets));
				}

				if(format == 'plaintext') {
					return res.attachment(board.name + '.txt').send(200, exportAs.generatePlainText(board, tickets));
				}
	
				if(format == 'image') { 
					return exportAs.postImage(req, board, tickets, function(options) {
						request.post(options).pipe(res);
					});
				}

				var boardObject     	= board;
				    boardObject.tickets = tickets;

				return res.attachment(board.name + '.json').json(200, boardObject);		
			});
		});
	});

Router.route('/boards/:board_id/tickets')

	/**
	 * Get the tickets belonging to the specified board.
	 *
	 * returns:
	 *   An array of ticket objects.
	 */
	.get(middleware.authenticate('user', 'guest'))
	.get(middleware.relation('admin', 'user', 'guest'))
	.get(function(req, res) {
		var board = req.resolved.board;
		Ticket.find({ 'board': board.id }, function(err, tickets) {
			if(err) {
				return next(utils.error(500, err));
			}
			return res.json(200, tickets);
		});
	})

	/**
	 * Create a new ticket on the specified board.
	 *
	 * TODO Validate the payload. Can be done in the 'ticket' model. Remember
	 *      to check for a 'ValidationError' on save.
	 *
	 * payload:
	 *   {
	 *     'color':   '#BABABA'
	 *     'heading': 'for sharks only'
	 *     'content': 'important stuff'
	 *     'position': {
	 *       'x', 'y', 'z'
	 *     }
	 *   }
	 *
	 * returns:
	 *   The created ticket object.
	 */
	.post(middleware.authenticate('user', 'guest'))
	.post(middleware.relation('admin', 'user', 'guest'))
	.post(function(req, res, next) {

		var payload       = req.body;
		    payload.board = req.resolved.board.id;

		new Ticket(payload).save(function(err, ticket) {
			if(err) {
				return next(utils.error(400, err));
			}

			new Event({
				'type': 'TICKET_CREATE',
				'board': ticket.board,
				'user': {
					'id':       req.user.id,
					'type':     req.user.type,
					'username': req.user.username,
				},
				'data': {
					'id':       ticket._id,
					'color':    ticket.color,
					'content':  ticket.content,
					'position': ticket.position,
				}
			}).save(function(err, ev) {
				if(err) {
					return console.error(err);
				}
				utils.emitter.to(ticket.board)
					.emit('board:event', ev.toObject());
			});

			/**
			 * Deprecated.
			 */
			utils.emitter.to(req.resolved.board.id)
				.emit('ticket:create', {
					user:   req.user,
					board:  req.resolved.board.id,
					ticket: ticket.toObject()
				});

			return res.json(201, ticket);
		});
	});


Router.route('/boards/:board_id/tickets/:ticket_id')

	/**
	 * Updates the given ticket.
	 *
	 * TODO Validate the payload.
	 *
	 * payload:
	 *   {
	 *     'color':    '#FFF'
	 *     'heading':  'new-heading'
	 *     'content':  'new-content'
	 *     'position': {
	 *       'x', 'y', 'z'
	 *     }
	 *   }
	 *
	 * returns:
	 *   The updated ticket object.
	 */
	.put(middleware.authenticate('user', 'guest'))
	.put(middleware.relation('admin', 'user', 'guest'))
	.put(function(req, res, next) {
		var old             = req.resolved.ticket.toObject();
		req.resolved.ticket = _.merge(req.resolved.ticket, req.body);

		return req.resolved.ticket.save(function(err, ticket) {
			if(err) {
				return next(utils.error(500, err));
			}

			if(!ticket) return next(utils.error(404, 'Ticket not found'));

			new Event({
				'type': 'TICKET_EDIT',
				'board': ticket.board,
				'user': {
					'id':       req.user.id,
					'type':     req.user.type,
					'username': req.user.username,
				},
				'data': {
					'id': ticket._id,

					'oldAttributes': {
						'color':    old.color,
						'heading':  old.heading,
						'content':  old.content,
						'position': old.position,
					},

					'newAttributes': {
						'color':    ticket.color,
						'heading':  ticket.heading,
						'content':  ticket.content,
						'position': ticket.position,
					},
				}
			}).save(function(err, ev) {
				if(err) {
					return console.error(err);
				}
				utils.emitter.to(ev.board)
					.emit('board:event', ev.toObject());
			});

			/**
			 * Deprecated.
			 */
			utils.emitter.to(req.resolved.board.id)
				.emit('ticket:update', {
					user:   req.user,
					board:  req.resolved.board.id,
					ticket: ticket.toObject()
				});

			return res.json(200, ticket);
		});
	})

	/**
	 * Delete the specified ticket.
	 *
	 * returns:
	 *   The deleted ticket object.
	 */
	.delete(middleware.authenticate('user', 'guest'))
	.delete(middleware.relation('admin', 'user', 'guest'))
	.delete(function(req, res, next) {
		req.resolved.ticket.remove(function(err) {
			if(err) {
				return next(utils.error(500, err));
			}

			new Event({
				'type': 'TICKET_REMOVE',
				'board': req.resolved.ticket.board,
				'user': {
					'id':       req.user.id,
					'type':     req.user.type,
					'username': req.user.username,
				},
				'data': {
					'id': req.resolved.ticket._id,
				}
			}).save(function(err, ev) {
				if(err) {
					return console.error(err);
				}
				utils.emitter.to(ev.board)
					.emit('board:event', ev.toObject());
			});

			/**
			 * Deprecated.
			 */
			utils.emitter.to(req.resolved.board.id)
				.emit('ticket:remove', {
					user:   req.user,
					board:  req.resolved.board.id,
					ticket: req.resolved.ticket.toObject()
				});

			return res.json(200, req.resolved.ticket);
		});
	});


Router.route('/boards/:board_id/tickets/:ticket_id/comments')

/**
 * Post a new comment on the specified ticket.
 */
	.post(middleware.authenticate('user', 'guest'))
	.post(middleware.relation('user', 'guest'))
	.post(function(req, res, next) {

		var old             = req.resolved.ticket.toObject();
		req.resolved.ticket = _.merge(req.resolved.ticket, req.body);

		var userId = null;

		// Backwards compatibility for older clients for guest users
		if(ObjectId.isValid(req.user.id)) {
			userId = req.user.id
		}

		req.resolved.ticket.comments.unshift({ 'user': { 'id': userId,
			'username': req.user.username},
			'content': req.body.comment});

		req.resolved.ticket.save(function (err, ticket) {
			if(err) {
				return next(utils.error(500, err));
			}

			if (!ticket) {
				return next(utils.error(404, 'Ticket not found'));
			}
			/*
			 * TODO: TICKET_COMMENT edit?
			 */
			new Event({
				'type': 'TICKET_EDIT',
				'board': ticket.board,
				'user': {
					'id':       req.user.id,
					'type':     req.user.type,
					'username': req.user.username,
				},
				'data': {
					'id': ticket._id,

					'oldAttributes': {
						'color':    old.color,
						'content':  old.content,
						'heading':  old.heading,
						'position': old.position,
						'comments': old.comments
					},

					'newAttributes': {
						'color':    ticket.color,
						'content':  ticket.content,
						'heading':  ticket.heading,
						'position': ticket.position,
						'comments': ticket.comments
					},
				}
			}).save(function(err, ev) {
					if(err) {
						return console.error(err);
					}
					utils.emitter.to(ev.board)
						.emit('board:event', ev.toObject());
				});
			return res.json(200, ticket);
		})
	});

Router.route('/boards/:board_id/events')

	/**
	 * Get a list of 'events' that have occurred on this board.
	 *
	 * returns:
	 *   [ EventObject ]
	 */
	.get(middleware.authenticate('user', 'guest'))
	.get(middleware.relation('admin', 'user', 'guest'))
	.get(function(req, res, next) {
		Event.find({ 'board': req.resolved.board.id }, function(err, evs) {
			if(err) {
				return next(utils.error(500, err));
			}
			return res.json(200, evs);
		});
	});


Router.route('/boards/:board_id/access')

	/**
	 * Get the 'access-code' of the given board.
	 *
	 * returns:
	 *   {
	 *     'accessCode': 'WxYz8Bd'
	 *   }
	 */
	.get(middleware.authenticate('user'))
	.get(middleware.relation('admin'))
	.get(function(req, res) {
		return res.json(200, {
			accessCode: req.resolved.board.accessCode || ''
		});
	})

	/**
	 * Generate a new 'access-code' for the given board. Will replace any
	 * existing 'access-code', invalidating any 'guest-tokens' tied to it.
	 *
	 * returns:
	 *   {
	 *     'accessCode': 'WxYz8Bd'
	 *   }
	 */
	.post(middleware.authenticate('user'))
	.post(middleware.relation('admin'))
	.post(function(req, res, next) {
		// Generate an 'access-code' for the specified board.
		var accessCode = require('crypto').randomBytes(4).toString('hex');

		// Attach the generated 'access-code' to the specified board.
		req.resolved.board.accessCode = accessCode;
		req.resolved.board.save(function(err, board) {
			if(err) {
				return next(utils.error(500, err));
			}

			new Event({
				'type': 'BOARD_PUBLISH',
				'board': req.resolved.board.id,
				'user': {
					'id':       req.user.id,
					'type':     req.user.type,
					'username': req.user.username,
				},
				'data': {
					'accessCode': accessCode,
				}
			}).save(function(err, ev) {
				if(err) {
					return console.error(err);
				}
				utils.emitter.to(ev.board)
					.emit('board:event', ev.toObject());
			});

			return res.json(200, { accessCode: board.accessCode });
		});
	})

	/**
	 * Remove the existing 'access-code' invalidating all 'guest-tokens' tied
	 * to it. This will essentially hide the board from outside eyes.
	 */
	.delete(middleware.authenticate('user'))
	.delete(middleware.relation('admin'))
	.delete(function(req, res, next) {
		req.resolved.board.accessCode = null;
		req.resolved.board.save(function(err) {
			if(err) {
				return next(utils.error(500, err));
			}

			new Event({
				'type': 'BOARD_UNPUBLISH',
				'board': req.resolved.board.id,
				'user': {
					'id':       req.user.id,
					'type':     req.user.type,
					'username': req.user.username,
				},
			}).save(function(err, ev) {
				if(err) {
					return console.error(err);
				}
				utils.emitter.to(ev.board)
					.emit('board:event', ev.toObject());
			});

			return res.send(200);
		});
	});

Router.route('/boards/:board_id/access/:code/grantaccess')

/**
 * Gives a user access to a board
 *
 */
	.put(middleware.authenticate('user'))
	.put(function(req, res, next) {

		var board = req.resolved.board;
		var user = req.user;
		var isMember = false;
		// User is already a member of this  baord.
		board.members.map(function (member) {
			if (member.id == user.id) {
				isMember = true;
			}
		});

		if(!isMember) {
			board.members.push({id: user.id, role: 'member'});
			board.save(function (err, board) {
				if (err) {
					return next(utils.error(500, err));
				}

				User.findOne({_id: user.id}, function (err, doc) {
					doc.boards.push({id: board._id});
					doc.save(function (err) {
						if (err) return console.error(err);

						return res.json(200, board);
					});
				});
			});
		}
		else {
			return res.json(200, board);
		}
	});


Router.route('/boards/:board_id/access/:code')

	/**
	 * Generates a 'guest' token to the given board.
	 *
	 * payload:
	 *   {
	 *     'username': 'nick'
	 *   }
	 *
	 * sets headers:
	 *   'x-access-token' : 'guest-token'
	 */
	.post(function(req, res, next) {
		var jwt   = require('jsonwebtoken');
		var board = req.resolved.board;

		// Requested board must have a 'accessCode' set.
		if(!board.accessCode || board.accessCode != req.params.code) {
			return next(utils.error(401, ''));
		}

		// TODO Guest must have a valid 'username'.
		var guestPayload = {
			id:         require('crypto').randomBytes(4).toString('hex'),
			type:       'temporary',
			access:     board.id,
			username:   req.body.username,
			accessCode: board.accessCode
		}

		// Generate the 'guest-token' for access.
		var guestToken = jwt.sign(guestPayload, config.token.secret);

		var session = {
			user_agent: req.headers['user-agent'],
			token:      guestToken,
			created_at: new Date()
		};

		new User({ name: req.body.username,
			account_type: 'temporary',
			created_at: new Date(),
			sessions: [session] })
			.save(function(err, user) {
				if(err) {
					if(err.name == 'ValidationError') {
						return next(utils.error(400, err));
					}
					if(err.name == 'MongoError' && err.code == 11000) {
						return next(utils.error(409, 'User already exists'));
					}
					return next(utils.error(500, err));
				}
			});


		new Event({
			'type': 'BOARD_GUEST_JOIN',
			'board': req.resolved.board.id,
			'user': {
				'id':       guestPayload.id,
				'type':     guestPayload.type,
				'username': guestPayload.username,
			},
		}).save(function(err, ev) {
			if(err) {
				return console.error(err);
			}
			utils.emitter.to(ev.board)
				.emit('board:event', ev.toObject());
		});

		return res.set('x-access-token', guestToken)
			.json(200, guestPayload);
	});

module.exports = Router;
